### PR TITLE
[WIP] Update Yellow Highlight Dismissal Behavior

### PR DESCRIPTION
**Pull request by brunovjk**
### Details

This update modifies the behavior of the yellow highlight for linked report actions to dismiss under various conditions.

### Fixed Issues

$ https://github.com/Expensify/App/issues/42165
PROPOSAL: https://github.com/Expensify/App/issues/42165#issuecomment-2111017989

### Tests

1. Click into a thread
2. Click back into the main conversation by clicking the the room title in the top left of the thread window
3. Observe that the thread is highlighted in yellow after you exit it
4. **The yellow highlight should fade after:**
   1. Clicking out of the chat into another chat
   2. Clicking into another thread
   3. Refreshing the page
   4. Sending a new comment in the same chat (but not receiving a comment from another user)
   5. Taking any action from the + menu while in the chat

- [x] Verify that no errors appear in the JS console

### Offline tests

Same as Tests

### QA Steps

Same as Tests

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/d12a1eb4-79f2-40eb-991b-4a2e0cc16d8c

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/d501f2b4-ab2b-469c-aa82-6a434d42b5b2

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/0f27cd91-4982-44d7-84de-306a9623691d

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/eb7472f7-4ae0-4b6d-a443-6a495cb8f27f

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/32e37331-4050-443a-97ca-c025cb324cea

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/9cd326a8-2e24-433e-bf16-355560bf6b82

</details>

### CODE REVIEW

File: src/pages/home/report/ReportActionsList.tsx [hunk 1]
```diff
@@ -341,13 +355,20 @@ function ReportActionsList({
...
                 return;
             }
-            InteractionManager.runAfterInteractions(() => reportScrollManager.scrollToBottom());
+            InteractionManager.runAfterInteractions(() => {
+                // Reset highlight if necessary
+                if (linkedReportActionID) {
+                    Navigation.setParams({reportActionID: ''});
+                }
+                reportScrollManager.scrollToBottom();
+            });
```
Comment (brunovjk) [inline src/pages/home/report/ReportActionsList.tsx:369]: @sobitneupane: If `!hasNewestReportActionRef.current` we don't reset the highlight or scroll to the bottom.
In a chat with many new messages, when the current user creates a new action, the scroll doesn't go to the bottom, consequently the yellow highlight isn't cleared.\
Should we leave it like that, or should we clear the yellow highlight regardless of `hasNewestReportActionRef.current`? What do you think?

File: src/pages/home/report/ReportActionsList.tsx [hunk 2]
```diff
@@ -329,6 +330,19 @@ function ReportActionsList({
...
+            return;
+        }
+        const navigationEntries = performance.getEntriesByType('navigation');
+        if (!navigationEntries.length || (navigationEntries[0] as PerformanceNavigationTiming).type !== 'reload') {
+            return;
+        }
+        Navigation.setParams({reportActionID: ''});
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [linkedReportActionID]);
+
```
Comment (brunovjk) [inline src/pages/home/report/ReportActionsList.tsx:345]: @sobitneupane I tested some alternatives and didn't find a better way to **identify when the user refreshed the page** without breaking the "Comment Linking" feat. Can you think of a better way?

**Comment by brunovjk** [comment]
@sobitneupane I think found a bug, not related to this PR, however making it not possible to test correctly:
Only in `native`, when we return to the main chat, the yellow highlight appears as expected, but the new message button appears no matter the size of the conversation. Even if the user takes action, the scroll doesn't go down, and the yellow highlight doesn't disappear. I did a quick search and didn't find a related issue, I'll look further.

File: src/pages/home/report/ReportActionsList.tsx (hunk 2 above)
Comment (brunovjk) [reply src/pages/home/report/ReportActionsList.tsx:345]: @sobitneupane I think found a bug, not related to this PR, however making it not possible to test correctly:
Only in `native`, when we return to the main chat, the yellow highlight appears as expected, but the new message button appears no matter the size of the conversation. Even if the user takes action, the scroll doesn't go down, and the yellow highlight doesn't disappear. I did a quick search and didn't find a related issue, I'll look further.

File: src/pages/home/report/ReportActionsList.tsx (hunk 1 above)
Comment (sobitneupane) [reply src/pages/home/report/ReportActionsList.tsx:369]: @brunovjk I believe we should clear the highlight if any new action is taken by the user.

File: src/pages/home/report/ReportActionsList.tsx (hunk 2 above)
Comment (sobitneupane) [reply src/pages/home/report/ReportActionsList.tsx:345]: @brunovjk Can you please share the screen recording of the issue. 

The issue is "We don't remove the yellow highlight and scroll  doesn't go down even after user takes action in native devices", right? Is [this](https://github.com/Expensify/App/pull/43581#discussion_r1645019146) the same issue?

Comment (brunovjk) [reply src/pages/home/report/ReportActionsList.tsx]: **Reproduced in the last main 06/20**:
<details>
<summary>When returning from a theard, the "New Message" flag is active even when we have few messages, and does not disappear when scrolling.</summary>



https://github.com/Expensify/App/assets/95647348/57da200e-3f1f-4a5f-bc14-0c5c42c28c8a

</details>

Steps to reproduce:
1. Login to the native app (iOS or Android)
2. Go to any chat, click on "Reply on Theard" in the last message of that chat.
3. Click back into the main conversation by clicking the the room title in the top left of the thread window
4. Thread is highlighted in yellow after you exit it
5. Note that the "New Message" flag is active even when we have few messages, and does not disappear when scrolling. Try send a new message, it also doesn't desappear and it doesn't scroll to the end. On other platforms it works normally.

File: src/pages/home/report/ReportActionsList.tsx (hunk 1 above)
Comment (brunovjk) [reply src/pages/home/report/ReportActionsList.tsx]: > @brunovjk I believe we should clear the highlight if any new action is taken by the user.

Exactly, I tested calling `clearHighlight` as soon as the user performs an action, but the application looks strange and renders more than necessary, blinking you know.

I tried several different flows/steps to call `clearHighlight`, the best way I found was to call it in `runAfterInteractions` here:
https://github.com/Expensify/App/blob/896c14c96333c34a43a82191a7170333dddd49cf/src/pages/home/report/ReportActionsList.tsx#L349

but for it to work well, everything before, like this check:
https://github.com/Expensify/App/blob/896c14c96333c34a43a82191a7170333dddd49cf/src/pages/home/report/ReportActionsList.tsx#L346
should work too. I believe they are trying to resolve it as we speak: https://github.com/Expensify/App/pull/39951#discussion_r1647084530

File: src/pages/home/report/ReportActionsList.tsx (hunk 2 above)
Comment (sobitneupane) [reply src/pages/home/report/ReportActionsList.tsx]: @brunovjk Has this been resolved?

Comment (brunovjk) [reply src/pages/home/report/ReportActionsList.tsx]: <details>
<summary>Not yet, I haven't found any related issue, I'll report it.</summary>


https://github.com/Expensify/App/assets/95647348/ffcd1e0d-f0dc-4e02-a15d-cb3ee760ff42

</details>

Anyway, I'm still not 100% confident how Im dealing with "checking if the user refreshed the page". I'm working on it at the moment.

Comment (brunovjk) [reply src/pages/home/report/ReportActionsList.tsx]: > @brunovjk Has this been resolved?

@ sobitneupane The issue has been resolved. I'll keep testing :D

Comment (sobitneupane) [reply src/pages/home/report/ReportActionsList.tsx]: Great. Please let me know when the PR is ready for review.

**Comment by brunovjk** [comment]
Hello @sobitneupane, I need some help.

So far our _"Yellow Highlight Dismissal Behavior"_ works as expected for the following scenarios:
- [x] Clicking out of the chat into another chat
- [x] Clicking into another thread
- [ ] Refreshing the page
- [x] Sending a new comment in the same chat (but not receiving a comment from another user)
- [x] Taking any action from the + menu while in the chat

However, I’m having **trouble** with **dismissing the highlight after refreshing the page**. I've tried using window/performance detection, useState/useEffect logic, and the beforeunload event, but nothing seems to work perfectly.

Could you give me some suggestions on how can we achieve this? Thank you very much.

**Comment by sobitneupane** [comment]
Sorry for the delay @brunovjk. I running quite busy. I will try to get to it asap.

**Comment by brunovjk** [comment]
Okay, I think I found a safe way, I'm testing it and I'll let you know. Thanks for your patience.

**Comment by brunovjk** [comment]
@sobitneupane PR, finally, ready for review. Looking forward to hearing your thoughts on the hook I created.

**Comment by sobitneupane** [comment]
> Taking any action from the + menu while in the chat

@brunovjk Should we remove the yellow highlight as soon as we begin taking any action (such as Submit expense), or after the action is complete? I am inclined towards the former.

**Comment by brunovjk** [comment]
Cool, let me change it now. One moment

**Comment by brunovjk** [comment]
@sobitneupane now available for review. Thank you very much.

**Comment by sobitneupane** [comment]
@brunovjk I can still reproduce [this issue](https://github.com/Expensify/App/pull/43581#issuecomment-2271042767).

https://github.com/user-attachments/assets/8f2e6998-8f52-4e16-878b-c158e197d506

**Comment by brunovjk** [comment]
Thank you, Im on it

**Comment by brunovjk** [comment]
>  Screen.Recording.2024-08-16.at.14.00.30.mov

@sobitneupane Thank you for clarifying. I misunderstood your point about _"removing as soon as we start taking any action."_ It's more accurate now. However, when cleaning the `reportActionID` (which also clears the yellow highlight) by starting an action in a chat with many messages, the highlighted item, which might be well above the visible area, causes the view to scroll to the bottom:

https://github.com/user-attachments/assets/832f7249-f188-464a-8567-9b7309671581

What do you think? Should we keep the current behavior of scrolling to the bottom, or should we only remove the highlight while keeping the `reportActionItem` visible? Thanks :D

**Comment by sobitneupane** [comment]
> What do you think? Should we keep the current behavior of scrolling to the bottom, or should we only remove the highlight while keeping the reportActionItem visible?

@brunovjk I think we should only remove the highlight.

**Comment by brunovjk** [comment]
> @brunovjk Any update? Let's try to get this merged asap.

Still, there's just one detail missing now. In short: The yellow highlight is linked to the `linkedReportActionId`, we must 'clear the route parameters' with this function I created:
https://github.com/Expensify/App/blob/67ff90351f41be0666f79d8aae575b02c1c563da/src/libs/actions/Report.ts#L441-L450

But we only have one problem now, if the highlighted message is in a chat with many messages below it, when clearing the `reportActionID` we do scroll down, as indicated [here](https://github.com/Expensify/App/pull/43581#issuecomment-2297263876).

I'm focused on this, it should have been closed a long time ago, but do you have any idea how to prevent scrolling when we 'clear' the `reportActionID`?? Thanks.

**Comment by brunovjk** [comment]
Making good progress here, I need to do some more testing to make sure everything is ok and no regressions, and we'll have it for review soon.

**Comment by brunovjk** [comment]
Misclicked above.

**Comment by brunovjk** [comment]
Hey @sobitneupane :D could you please take a look at the changes I made? Everything looks good to me now.

**Comment by sobitneupane** [comment]
> Edit: I'm investigating the Reassure Performance Tests fail, do you have any idea what might be causing it?

I am not sure. @brunovjk Is it related to this PR? Which part of the code do you think is causing the issue?

**Comment by brunovjk** [comment]
@sobitneupane could you take a look at the changes now? I tested it and everything seems to work fine, but I'm still a little unsure about the names of the variables and functions we created.

**Comment by sobitneupane** [comment]
Thanks for the update @brunovjk

**Comment by sobitneupane** [comment]
> Refreshing the page

@brunovjk I can still reproduce the issue. The highlight doesn't fade after refreshing the page.


https://github.com/user-attachments/assets/70eeaeff-1951-44a0-b175-69600d04fe89

**Comment by brunovjk** [comment]
**Update**: Working on fixing/updating [the hook](https://github.com/brunovjk/Expensify-App/blob/fix/42165/src/hooks/usePageRefresh.ts) to a more robust approach, I'll update again by the end of the week.

**Pull request closed without merging**